thunder/transforms/autodiff.py

```diff
@@ -0,0 +1,311 @@
+                                    grad_proxy_map[name] = new_grad
+                                # elif KNOWN proxy:
+                                #    ### TODO: it can happen that an intermediate does not have put_grad, e.g. first arg to where
+                                else:
+                                    raise RuntimeError(f"grad of non-output and non-intermediate {name} requested")
+                            else:
+                                self.add_processed_bsyms([nbsym.from_bsym()])
+                self.collected_bw_part_bsyms.clear()
+
+                grad_flat_args = []
+                for p in bsym.args[0]["flat_args"]:
+                    # or p = self.read(p) here?
+                    if isinstance(p, thunder.Proxy) and p.requires_grad and p.name in grad_proxy_map:
+                        # is it always OK if we don't have a gradient? (one case: unused input)
+                        # result of put_grad???
+                        grad_flat_args.append(grad_proxy_map[p.name])
+                    else:
+                        grad_flat_args.append(None)
+
+                bsym.args[0]["grad_flat_args"] = grad_flat_args
+                self.add_processed_bsyms([bsym.from_bsym()])
+                self.set_result(bsym.output)
+                return
+
+            # constant for gradients (no grad required)
+            if is_constant_for_vjp(bsym):
+                if bsym.sym.name == "synchronize":
+                    # This is a *really* terrible hack to cope with non-grad-needing sharded tensors
+                    # as required by LoRA.
+                    from thunder.distributed.prims import all_gather
+
+                    def synchronize_impl(a, group):
+                        return all_gather(a, group, True).wait()
+
+                    self.add_bsyms_from_function(synchronize_impl, *bsym.args, **bsym.kwargs)
+                    return
+
+                self.add_processed_bsyms([bsym.from_bsym()])
+                self.set_result(bsym.output)
+                return
+
+            # executor or global grad transform
+            joint_forward_backward, _ = _get_gradfn_and_executor(bsym)
+
+            if joint_forward_backward is None:
+                # registered augmented forward impl (where aug fwd and backward are registered separately)
+                aug_fwd_impl = augmented_forward_impls.get(bsym.sym.id)
+                if aug_fwd_impl is not None:
+                    bwd_impl = backward_impls.get(bsym.sym.id)
+
+                    def joint_forward_backward(*args, **kwargs):
+                        aug_fwd_res = aug_fwd_impl(*bsym.args, **bsym.kwargs)
+                        if isinstance(aug_fwd_res, thunder.core.transforms.VJPDual):
+                            res, saved_for_backward = aug_fwd_res
+                        else:
+                            res, saved_for_backward = aug_fwd_res
```

Comment:
This looks a little funny.

Reply:
I need to figure out what the return convention for aug forward actually is.

Also for backward:

https://github.com/Lightning-AI/lightning-thunder/blob/150d7e0bebda9c2f27bddcf7e21206d0c7d37ee9/thunder/core/transforms.py#L2274-L2280

why is it not `return None, g_weight`

And not to mention that we also have this:

https://github.com/Lightning-AI/lightning-thunder/blob/150d7e0bebda9c2f27bddcf7e21206d0c7d37ee9/thunder/core/transforms.py#L1360-L1377

which looks fishy to me (shouldn't the `put_grad` be on `weight` and `a` is the index?).

Reply:
fixed now.